### PR TITLE
feat: Elastic energy

### PR DESCRIPTION
**Pull request by Vollkornaffe**
This adds functions to compute the elastic energy density function from a given deformation gradient.

The new code is different from the existing one.
See `pos_energy`: https://github.com/dimforge/sparkl/blob/43c13a65fef40b7872ec638c866c4b42926f3354/src/dynamics/models/elasticity_corotated_linear.rs#L35-L37
We may want to refactor.

### CODE REVIEW

File: src/dynamics/models/elasticity_corotated_linear.rs [hunk 1]
```diff
@@ -17,6 +17,21 @@ impl ConstitutiveModel for CorotatedLinearElasticity {
         )
     }
 
+    // https://www.math.ucla.edu/~cffjiang/research/mpmcourse/mpmcourse.pdf#subsection.6.3
```
Comment (sebcrozet) [inline src/dynamics/models/elasticity_corotated_linear.rs]:
```suggestion
    // https://www.math.ucla.edu/~cffjiang/research/mpmcourse/mpmcourse.pdf#subsection.6.3
    // This is essentially the same calculation as `[pos_energy]` but without ignoring the negative part.
```

File: src/dynamics/models/elasticity_corotated_linear.rs [hunk 2]
```diff
@@ -17,6 +17,21 @@ impl ConstitutiveModel for CorotatedLinearElasticity {
...
+            .svd_unordered(false, false)
+            .singular_values;
+        let determinant: Real = singular_values.iter().product();
+
+        self.mu
+            * singular_values
+                .iter()
+                .map(|sigma| (sigma - 1.).powi(2))
+                .sum::<Real>()
+            + self.lambda / 2. * (determinant - 1.).powi(2)
```
Comment (sebcrozet) [inline src/dynamics/models/elasticity_corotated_linear.rs]: I would recommend taking the whole `&Particle` as argument, for consistency with `pos_energy` and to take hardening into account the same way as in `pos_energy`.

Can be in another PR.

File: src/dynamics/models/elasticity_neo_hookean.rs [hunk 3]
```diff
@@ -17,6 +17,15 @@ impl ConstitutiveModel for NeoHookeanElasticity {
...
 
+    // https://www.math.ucla.edu/~cffjiang/research/mpmcourse/mpmcourse.pdf#subsection.6.2
+    fn elastic_energy_density(&self, deformation_gradient: Matrix<Real>) -> Real {
+        let determinant_log = deformation_gradient.determinant().ln();
+        self.mu / 2.
+            * ((deformation_gradient.transpose() * deformation_gradient).trace() - DIM as Real)
+            - self.mu * determinant_log
+            + self.lambda / 2. * determinant_log.powi(2)
+    }
+
```
Comment (sebcrozet) [inline src/dynamics/models/elasticity_neo_hookean.rs]: Same remark regarding taking hardening into account.

File: src_core/dynamics/models/elasticity_corotated_linear.rs [hunk 4]
```diff
@@ -76,7 +76,7 @@ impl CorotatedLinearElasticity {
 
     pub fn pos_energy(&self, deformation_gradient: Matrix<Real>, elastic_hardening: Real) -> Real {
         let j = deformation_gradient.determinant();
-        let mut pos_def = deformation_gradient.svd_unordered(true, true);
+        let mut pos_def = deformation_gradient.svd_unordered(true, true); // TODO: why compute U and V?
```
Comment (sebcrozet) [inline src_core/dynamics/models/elasticity_corotated_linear.rs]: Probably just a refactoring leftover. It can be `false, false`.

**Review by sebcrozet — APPROVED** [review]

File: src_core/dynamics/models/elasticity_neo_hookean.rs [hunk 5]
```diff
@@ -75,6 +76,48 @@ impl NeoHookeanElasticity {
         false
     }
 
+    // General elastic density function: https://dl.acm.org/doi/pdf/10.1145/3306346.3322949#section.5 (8) and (9)
+    // With degradation: https://dl.acm.org/doi/pdf/10.1145/3306346.3322949#subsection.3.2 (3)
+    // With hardening: https://www.math.ucla.edu/~cffjiang/research/mpmcourse/mpmcourse.pdf#subsection.6.5 (87)
+    #[allow(non_snake_case)]
```
Comment (sebcrozet) [inline src_core/dynamics/models/elasticity_neo_hookean.rs]: Let’s not allow `non_snake_case`. It messes up with syntax highlighting, and is very unconventional in rust, even though I agree it makes it closer to the mathematical notation.

**Review by sebcrozet — CHANGES_REQUESTED** [review]

File: src_core/dynamics/models/elasticity_neo_hookean.rs [hunk 6]
```diff
@@ -75,6 +76,48 @@ impl NeoHookeanElasticity {
...
+
+        let J = F.determinant();
+
+        let alpha = -1. / DIM as Real;
+
+        let J_alpha = J.powf(alpha);
+
+        let Psi_mu =
+            |F: Matrix<Real>| hardened_mu / 2. * ((F.transpose() * F).trace() - DIM as Real);
+        let Psi_mu = Psi_mu(J_alpha * F);
```
Comment (sebcrozet) [inline src_core/dynamics/models/elasticity_neo_hookean.rs]: The closure makes the code more difficult to read, especially since it’s only called once and bound to a variable with the same name.

File: src/dynamics/models/elasticity_corotated_linear.rs (hunk 1 above)
Comment (Vollkornaffe) [reply src/dynamics/models/elasticity_corotated_linear.rs]: I added a reference to the `pos_energy` instead.

File: src_core/dynamics/models/elasticity_neo_hookean.rs (hunk 5 above)
Comment (Vollkornaffe) [reply src_core/dynamics/models/elasticity_neo_hookean.rs]: We're now using more descriptive names (with proper snake case) while keeping aliases matching the paper in the comments.

**Comment by Vollkornaffe** [comment]
Once we've converged on style for the neo-hookean one, I'll update the corotated code to match.

**Review by sebcrozet — APPROVED** [review]
Looks good now, thanks!
